cns/restserver/api.go

```diff
@@ -1506,3 +1507,56 @@ func (service *HTTPRestService) nmAgentSupportedApisHandler(w http.ResponseWrite
 
 	logger.Response(service.Name, nmAgentSupportedApisResponse, resp.ReturnCode, serviceErr)
 }
+
+func (service *HTTPRestService) registerNode(w http.ResponseWriter, r *http.Request) {
+	logger.Printf("[Azure CNS] registerNode")
+	logger.Request(service.Name, "registerNode", nil)
+	ctx := r.Context()
+
+	var (
+		returnCode    types.ResponseCode
+		req           cns.RegisterNodeRequest
+		returnMessage string
+		err           error
+	)
+
+	err = service.Listener.Decode(w, r, &req)
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		return
+	}
+
+	switch r.Method {
+	case http.MethodPost:
+		rnr := nma.RegisterNodeStandAloneRequest{
+			HomeAz: req.HomeAz,
+		}
+		err = service.nmagentMultiClient.NewClient.RegisterNodeStandAlone(ctx, rnr)
+		if err != nil {
+			apiError := nma.Error{}
+			if ok := errors.As(err, &apiError); ok {
+				switch apiError.StatusCode() {
+				case http.StatusInternalServerError:
+					returnMessage = "[Azure CNS] Error. registerNode failed due to Nmagent server internal error."
+					returnCode = types.NmAgentServerInternalError
+				case http.StatusBadRequest:
+					returnMessage = "[Azure CNS] Error. registerNode failed due to bad request error."
+					returnCode = types.NmAgentBadRequestError
+				default:
+					returnCode = types.UnexpectedError
+					returnMessage = fmt.Sprintf("[Azure CNS] Error. registerNode failed with StatusCode: %d", apiError.StatusCode())
+				}
```

Comment:
All of this information is encoded by the error itself. If you log the error, you will get something like:

```
nmagent: wireserver: http status 500: something terrible happened: body: a whole bunch of stuff
```

But it also might include issues with mime handling:

```
unexpected content type: application/xml: body: <some><weird><xml></xml></weird></some>
```

@rbtr Since these error codes are a public API, does adding these two make sense? Is there an existing one that's more appropriate? If not, would it make sense to add a code for the failure of a service dependency?

Reply:
@timraymond we briefly discussed this last week. any new idea on this? Thanks